06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+ - <i>numReconfigAttempts</i>: The value is incremented each time the node has been attempted to reconfigure.
+ - <i>numRetries</i>: The value is incremented each time the node cannot be restarted/reconfigured due to not meeting safety conditions (more on this later).
+ - <i>lastTransitionTime</i>: System.nanoTime of last observed state transition.
+
+ The following table illustrates the proposed node states and the possible transitions:
+ | State            | Description      | Possible transitions |
+ | :--------------- | :--------------- | :----------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running). This is determined via Kubernetes API, more details for it below. | `NOT_READY`, `NOT_RUNNING`. | `RESTARTED` `SERVING` | 
+ | NOT_READY             | Node is running but not ready to serve requests which is determined by Kubernetes readiness probe (broker state < 2 OR == 127 OR controller is not listening on port). | `RESTARTED` `SERVING` |
+ | RESTARTED             | After successful `kubectl delete pod`. | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | RECONFIGURED          | After successful Kafka node config update via Admin client. | `NOT_RUNNING` `NOT_READY` `RESTARTED` `SERVING` |
+ | RECOVERING            | Node has started but is in log recovery (broker state == 2). | `SERVING` |
+ | SERVING               | Node is in running state and ready to serve requests which is determined by Kubernetes readiness probe (broker state >= 3 AND != 127 OR controller is listening on port). | `RESTARTED` `RECONFIGURED` `LEADING_ALL_PREFERRED` |
+ | LEADING_ALL_PREFERRED | Node is in running state and leading all preferred replicas. | None
+
+The definitions of broker states can be found via the following link: [Broker States](https://github.com/apache/kafka/blob/3.7/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java).
+
+The definitions of the possible restart reasons can be found via the following link: [Restart Reasons](https://github.com/strimzi/strimzi-kafka-operator/blob/0.40.0/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/RestartReason.java)
+
+#### NOT_RUNNING state
+
+If one of the following is true, then node's state is `NOT_RUNNING:
+- no pod exists for this node
+- unable to get the `Pod Status` for the pod 
+- the pod has `Pending` status with `Unschedulable` reason
+- the pod has container status `ContainerStateWaiting` with `CrashLoopBackOff` or `ImagePullBackOff` reason
+If none of the above is true but the node is not ready, then its state would be `NOT_READY`.
+
+#### Flow diagram describing the overall flow of the states
+![The new roller flow](./images/06x-new-roller-flow.png)
```

Comment:
Some comments about the diagram of the FSM ...

1. Each state has to be unique and not duplicated.
2. If you are grouping states together, does it really mean they are different state? Or are they different state with different context? It looks weird to me.
3. Where you use "Action and Transition State" I think that: Action should be cleared, because it represents an output for the FSM? Is there a real output or just transition? You don't need "Transition State" because it's defined by the arrow you have.

My gut general feeling is that ... you are describing a very complex state machine and transitions in the "Algorithm" section but then the visualization of it is really weak.

Reply:
I meant the diagram to be more from high level and then broken down with more details in the Algorithm section. Given that, we may have to change various things around states and stuff, I will take this suggestion and recreate the diagram.

Reply:
Is the diagram changed after my comments? It seems almost the same. Also, what's "Serving" or "Waiting" or "Restarted" or "Reconfigured" ... they are not listed as states in the table above. 
I can't really match the table with the diagram, sorry :-(

Reply:
@ppatierno I have updated the diagram now. It's still more of high level flow showing the transitions. The possible states are listed but does not mean they are grouped together. Depending on the state, a different action taken and the possible actions are also listed. Do you think it's clearer now? Or do you think I should break down to each state mapping to the possible actions, instead of listing them together in a same bubble?

Reply:
well an FSM state machine should have single states (but I can understand your grouping of more than one here, so let's leave it for now). What I can't understand is what's the hexagon and it's content. They are not states, right? I can't see them in the table. So their presence confuses me.
Also you are duplicating circles for Not_Ready/Not_Running/Recovery: you have one with them and Ready, and one with them alone. I think you should split and having Ready on its own, as well as Leading_All_preferred.
Finally as an FSM state diagram it should not contain other stuff like "Start" or "Desired State Reached", you are describing an FSM, and those ones are not states.

Reply:
Thanks for the feedback @ppatierno, I made another update on the diagram. Please let me know what you think.

Reply:
Good to hear that it is in the right direction. I've taken the suggestion and made another update :). Thanks @ppatierno

Reply:
What does exactly "Iterate" mean on the orange arrows?

Reply:
that's repeating the process on node if the desired state is not reached.